docs/rag_config.md
## Architecture

The overall config management works like this:
![](config_architecture.png)

This means the best way to configure your RAG deployment is to modify the `config.yaml` file. You can find configuration examples in the [Config Cookbook](cookbook.md)

This file is then loaded to instanciate a `RAGConfig` object which is used by the `RAG` class.

## Changing the configuration

In the default configuration template ([`backend/config.yaml`](backend/config.yaml)) you will find this:

```yaml
# This is the LLM configuration (&LLMConfig is a yaml anchor to reference this block further down in the conf file)
LLMConfig: &LLMConfig

# By default we're using a GPT model deployed on Azure.
# You should be able to use any langchain BaseChatModel listed here:
# https://github.com/langchain-ai/langchain/blob/master/libs/community/langchain_community/chat_models/__init__.py
source: AzureChatOpenAI"

# This is a key-value map of the parameters that will be passed to the
# langchain chat model object when it's created.
# Looking at the AzureChatOpenAI source code (https://github.com/langchain-ai/langchain/blob/master/libs/community/langchain_community/chat_models/azure_openai.py),
# we input the following params:
source_config:
    openai_api_type: azure
    openai_api_key: {{ OPENAI_API_KEY }}
    openai_api_base: https://poc-genai-gpt4.openai.azure.com/
    openai_api_version: 2023-07-01-preview
    deployment_name: gpt4v

# While the params in source_config are specific to each model,
# temperature is implemented by all BaseChatModel classes in langchain.
# We input it outside the source_config block
temperature: 0.1

# ... Rest of the config ...
```

Let's say you want to use a gemini-pro on Vertex AI instead. [This is the source code of this model in langchain](https://github.com/langchain-ai/langchain/blob/master/libs/community/langchain_community/chat_models/vertexai.py#L206C7-L206C19):

```python
class ChatVertexAI(_VertexAICommon, BaseChatModel):
    """`Vertex AI` Chat large language models API."""

    model_name: str = "chat-bison"
    "Underlying model name."
    examples: Optional[List[BaseMessage]] = None

    ...
```

The  `config.yaml` could look like this:
```yaml
LLMConfig: &LLMConfig
  source: ChatVertexAI
  source_config:
    model_name: gemini-pro
  temperature: 0.1
```

Now, the next time you start the API, it will call gemini-pro instead of openAI.

## Extending the configuration

As you tune this starter kit to your needs, you may need to add specific configuration that your RAG will use.

For example, let's say you want to add the `foo` configuration parameter to your vector store configuration.

First, add it to `config.py` in the part relavant to the vector store:

```python
# ...

@dataclass
class VectorStoreConfig:
    # ... rest of the VectorStoreConfig ...

    foo: str = "bar"  # We add foo param, of type str, with the default value "bar"

# ...
```

This parameter will now be available in your `RAG` object configuration.

```python
from pathlib import Path
from backend.rag_components.rag import RAG

config_directory = Path("backend/config.yaml")
rag = RAG(config_directory)

print(rag.config.vector_store.foo)
# > bar
```

if you want to override its default value. You can do that in your `config.yaml`:
```yaml
VectorStoreConfig: &VectorStoreConfig
  # ... rest of the VectorStoreConfig ...
  foo: baz
```

```python
print(rag.config.vector_store.foo)
# > baz
```


docs/recipe_llms_configs.md
## Artefact Azure-hosted GPT4-turbo

```yaml
LLMConfig: &LLMConfig
  source: AzureChatOpenAI
  source_config:
    openai_api_type: azure
    openai_api_key: {{ OPENAI_API_KEY }}
    openai_api_base: https://genai-ds.openai.azure.com/
    openai_api_version: 2023-07-01-preview
    deployment_name: gpt4
  temperature: 0.1
```

## Local llama2
!!! info "You will first need to install and run Ollama"

    [Download the Ollama application here](https://ollama.ai/download)

    Ollama will automatically utilize the GPU on Apple devices.

    ```shell
    ollama run llama2
    ```

```yaml
LLMConfig: &LLMConfig
  source: ChatOllama
  source_config:
    model: llama2
```

## Vertex AI gemini-pro

!!! info "You will first need to login to GCP"

    ```shell
    export PROJECT_ID=<gcp_project_id>
    gcloud config set project $PROJECT_ID
    gcloud auth login
    gcloud auth application-default login
    ```

```yaml
LLMConfig: &LLMConfig
  source: ChatVertexAI
  source_config:
    model_name: gemini-pro
  temperature: 0.1
```


backend/config.py
import os
from dataclasses import dataclass, field, is_dataclass
from pathlib import Path

import yaml
from dotenv import load_dotenv
from jinja2 import Template
from langchain.chat_models.base import BaseChatModel
from langchain.schema.embeddings import Embeddings
from langchain.vectorstores import VectorStore

load_dotenv()

@dataclass
class LLMConfig:
    source: BaseChatModel | str = "AzureChatOpenAI"
    source_config: dict = field(default_factory=lambda: {
        "openai_api_type": "azure",
        "openai_api_base": "https://poc-genai-gpt4.openai.azure.com/",
        "openai_api_version": "2023-07-01-preview",
        "openai_api_key": os.environ.get("OPENAI_API_KEY"),
        "deployment_name": "gpt4v",
    })

    temperature: float = 0.1

@dataclass
class VectorStoreConfig:
    source: VectorStore | str = "Chroma"
    source_config: dict = field(default_factory=lambda: {
        "persist_directory": "vector_database/",
        "collection_metadata": {
            "hnsw:space": "cosine"
        }
    })

    retriever_search_type: str = "similarity"
    retriever_config: dict = field(default_factory=lambda: {
        "top_k": 20,
        "score_threshold": 0.5
    })

    insertion_mode: str = "full"  # "None", "full", "incremental"

@dataclass
class EmbeddingModelConfig:
    source: Embeddings | str = "OpenAIEmbeddings"
    source_config: dict = field(default_factory=lambda: {
        "openai_api_type": "azure",
        "openai_api_base": "https://poc-openai-artefact.openai.azure.com/",
        "openai_api_key": os.environ.get("EMBEDDING_API_KEY"),
        "deployment": "embeddings",
        "chunk_size": 500,
    })

@dataclass
class DatabaseConfig:
    database_url: str = os.environ.get("DATABASE_URL")

@dataclass
class RagConfig:
    llm:                      LLMConfig             = field(default_factory=LLMConfig)
    vector_store:             VectorStoreConfig     = field(default_factory=VectorStoreConfig)
    embedding_model:          EmbeddingModelConfig  = field(default_factory=EmbeddingModelConfig)
    database:                 DatabaseConfig        = field(default_factory=DatabaseConfig)
    chat_history_window_size: int                   = 5
    max_tokens_limit:         int                   = 3000
    response_mode:            str                   = None

    @classmethod
    def from_yaml(cls, yaml_path: Path, env: dict = None):
        if env is None:
            env = os.environ
        with Path.open(yaml_path, "r") as file:
            template = Template(file.read())
            config_data = yaml.safe_load(template.render(env))["RagConfig"]

        for field_name, field_type in cls.__annotations__.items():
            if field_name in config_data and is_dataclass(field_type):
                config_data[field_name] = field_type(**config_data[field_name])

        return cls(**config_data)
